### PR TITLE
Allow users to register names with greater granularity

### PR DESCRIPTION
**Pull request by talentlessguy**
this is how it looks:

image

mobile view:

image

image

extend UI (not bulk):

image

todo:

- [x] make calendar sync with year count
- [x] use old UI for bulk renewals
- [x] write tests for the one with calendar and perhaps mock `showPicker`

### CODE REVIEW

File: public/locales/zh/register.json [hunk 1]
```diff
@@ -3,7 +3,7 @@
   "subtitle": "注册",
   "heading": "注册 {{name}}",
   "invoice": {
-    "yearRegistration": "注册 {{years}} 年",
+    "yearRegistration": "注册 {{time}} 年",
```
Comment (talentlessguy) [inline public/locales/zh/register.json:6]: i'm pretty sure locale here is incorrect but idk what to change it to

File: src/utils/date.ts [hunk 2]
```diff
@@ -1,7 +1,25 @@
...
+  const month = String(date.getMonth() + 1).padStart(2, '0') // Month is zero-indexed
+  const day = String(date.getDate()).padStart(2, '0')
+  return `${year}-${month}-${day}`
+}
+
+export const getSecondsFromDate = (date: Date) => {
+  // @ts-ignore typescript doesn't support date operators
+  const value = Math.floor((date - now) / 1000)
+  return value
+}
```
Comment (talentlessguy) [inline src/utils/date.ts:25]: given that this function is only used once, probably it's better to just use a literal expr at where it's used?

File: src/components/pages/profile/[name]/registration/steps/Pricing/Pricing.tsx [hunk 3]
```diff
@@ -524,43 +529,45 @@ const Pricing = ({
...
-  const { hasPremium, premiumFee, gasPrice, yearlyFee, totalYearlyFee, estimatedGasFee } =
+
+  const { hasPremium, premiumFee, gasPrice, yearlyFee, totalDurationBasedFee, estimatedGasFee } =
     fullEstimate
+  const durationRequiredBalance = totalDurationBasedFee ? (totalDurationBasedFee * 110n) / 100n : 0n
+  const totalRequiredBalance = durationRequiredBalance
+    ? durationRequiredBalance + (premiumFee || 0n) + (estimatedGasFee || 0n)
+    : 0n
+
+  const previousYearlyFee = usePreviousDistinct(yearlyFee) || 0n
```
Comment (storywithoutend) [inline src/components/pages/profile/[name]/registration/steps/Pricing/Pricing.tsx:547]: What was the issue for why usePreviousDistinct was needed?

File: src/hooks/useRegistrationReducer.ts [hunk 4]
```diff
@@ -10,13 +10,14 @@ import {
   SelectedItemProperties,
 } from '@app/components/pages/profile/[name]/registration/types'
 import { useLocalStorageReducer } from '@app/hooks/useLocalStorage'
+import { yearsToSeconds } from '@app/utils/utils'
 
 const REGISTRATION_REDUCER_DATA_ITEM_VERSION = 2
```
Comment (storywithoutend) [inline src/hooks/useRegistrationReducer.ts:16]: Update this constant when making changes to registration data

File: src/utils/date.ts [hunk 5]
```diff
@@ -1,7 +1,25 @@
+const now = new Date()
+
```
Comment (storywithoutend) [inline src/utils/date.ts]: Wouldn't this keep the value constant after the file is loaded? Is probably safer to move to getSecondsFromDate

File: src/utils/utils.ts [hunk 6]
```diff
@@ -44,6 +48,25 @@ export const formatDateTime = (date: Date) => {
 export const formatFullExpiry = (expiryDate?: Date) =>
   expiryDate ? `${formatExpiry(expiryDate)}, ${formatDateTime(expiryDate)}` : ''
 
+export const formatExtensionPeriod = (duration: number) => {
+  const month = ONE_DAY * 30 // Assuming 30 days per month for simplicity
+
```
Comment (storywithoutend) [inline src/utils/utils.ts:53]: Make translation friendly

**Review by storywithoutend — DISMISSED** [review]

File: src/components/pages/profile/[name]/registration/steps/Pricing/Pricing.tsx [hunk 7]
```diff
@@ -470,6 +472,8 @@ export type PricingProps = {
   >['initiateMoonpayRegistrationMutation']
 }
 
+const minSeconds = ONE_YEAR
+
```
[the remark this answers is not in the document]
Comment (talentlessguy) [reply src/components/pages/profile/[name]/registration/steps/Pricing/Pricing.tsx:476]: from what I know the whole point of this PR was to allow extensions for less than a year mostly. I can allow registering for less than a year as well with a default of a year, but I'm not sure if that's the intention.

File: src/components/pages/profile/[name]/registration/steps/Pricing/Pricing.tsx (hunk 3 above)
Comment (talentlessguy) [reply src/components/pages/profile/[name]/registration/steps/Pricing/Pricing.tsx:547]: the gas preview was incorrect

File: src/components/@molecules/DateSelection/DateSelection.tsx [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [dateInYears, yearPickView])
+
+  return (
+    <Container>
+      {yearPickView === 'date' ? (
+        <Calendar
+          value={now + seconds}
+          onChange={(e) => {
+            const { valueAsDate } = e.currentTarget
```
[the remark this answers is not in the document]
Comment (talentlessguy) [reply src/components/@molecules/DateSelection/DateSelection.tsx:75]: do you mean unit tests? or e2e? e2e already covers this

File: src/utils/utils.ts [hunk 9]
```diff
@@ -15,15 +18,16 @@ export const shortenAddress = (address = '', maxLength = 10, leftSlice = 5, righ
...
-export const yearsToSeconds = (years: number) => years * 60 * 60 * 24 * 365
-
-export const secondsToYears = (seconds: number) => seconds / (60 * 60 * 24 * 365)
+export const deriveYearlyFee = ({
+  duration,
+  price,
+}: {
+  duration: number
+  price: { base: bigint }
+}) => {
```
[the remark this answers is not in the document]
Comment (talentlessguy) [reply src/utils/utils.ts:27]: image

not sure if that's the correct type

File: public/locales/en/transactionFlow.json [hunk 10]
```diff
@@ -221,7 +221,7 @@
         "description_one": "Extending this name will extend the current owner's registration length. This will not give you ownership of it.",
         "description_other": "Extending these names will extend the current owner's registration length. This will not give you ownership if you are not already the owner"      },
       "invoice": {
-        "extension": "{{count}} year extension",
```
[the remark this answers is not in the document]
Comment (talentlessguy) [reply public/locales/en/transactionFlow.json:224]: `formatExtensionPeriod` now returns localized strings

File: e2e/specs/stateless/extendNames.spec.ts [hunk 11]
```diff
@@ -310,3 +318,66 @@ test('should not show extend button on unwrapped subnames', async ({
...
+    const monthLater = await page.evaluate(
+      (ts) => {
+        return new Date(ts)
+      },
+      (expiryTime + daysToSeconds(30)) * 1000,
+    )
+
+    await calendar.fill(dateToDateInput(monthLater))
+    await expect(page.getByTestId('calendar-date')).toHaveValue(
+      secondsToDateInput(expiryTime + roundDurationWithDay(monthLater, expiryTime)),
```
Comment (talentlessguy) [inline e2e/specs/stateless/extendNames.spec.ts:363]: test passes locally but fails in CI. has something to do with time manipulation i suppose

**Review by sugh01 — APPROVED** [review]

**Pull request merged**